doc/cluster_design.md

```diff
@@ -0,0 +1,137 @@
+# Design: Clustering in SQLflow to analyze patterns in data
+
+## Concept
+
+For analysts and real business people, in the daily analysis work, most of the work is not prediction, but analysis of the patterns in the data. This can help them mine user behavioral characteristics and differences, helping the business discover value and operate.
+
+This design doc introduces how to support the `Cluster Model` in SQLFlow. 
+
+## User interface
+
+In this scenario, we focus on the extraction of data patterns in unsupervised learning. SO, we plan to use a **TRAIN SQL** to train a unsupervised model. We will support whether to perform Pretrain at the beginning of this unsupervised network in `WITH`, and whether to use the already trained model as a pre-training in `USING`. The simple pipeline like:
+
+
+TRAIN SQL:
+
+``` sql
+SELECT * FROM input_table
+TRAIN clusterModel
+WITH
+    model.encode_units = [100, 7]
+    model.n_clusters = 5
+    model.run_pretrain = false
+COLUMN m1, m2, m3, m4, m5, m6, m7, m8, m9, m10 
+USING model.existed_pretrain_model =  existed_pretrain_model
```

Comment:
In my opinion, it can be divided into such two scenarios:

1. User has no pre-train model (auto-encoder). 
 In this scenario the user wants the full training process which consists of `auto-encoder` and `clustering`. Thus the user should  be forbidden to use the `USING` clause, because there is no pre-train model ready for using in training process. User should define relate parameters of `auto-encoder` model  in `WITH` clause clearly, like `model.encode_units`.

2. User has pre-train model (auto-encoder).
In this case, use has at least one pre-train `auto-encoder` model already and he/she want to use it without training this part again. `USING` clause should be used for defining the path/name of the pre-train `auto-encoder` model and  in `WITH` clause user should guarantee the correct structure of pre-train model to make sure that the model data can be loaded correctly. It requires some additional checks to be performed in the background.